packages/markers/src/browser/problem/problem-tabbar-decorator.ts

```diff
@@ -49,9 +49,27 @@ export class ProblemTabBarDecorator implements TabBarDecorator {
         if (Navigatable.is(widget)) {
             const resourceUri = widget.getResourceUri();
             if (resourceUri) {
-                return this.problemManager.findMarkers({
-                    uri: resourceUri
-                }).map(marker => this.toDecorator(marker));
+                // Get the list of problem markers for the given resource URI.
+                const markers: Marker<Diagnostic>[] = this.problemManager.findMarkers({ uri: resourceUri });
+                // If no markers are available, return early.
+                if (markers.length === 0) {
+                    return [];
+                }
+                // Store the marker with the highest severity.
+                let maxSeverity: Marker<Diagnostic> | undefined;
+                // Iterate over available markers to determine that which has the highest severity.
+                // Only display a decoration if an error or warning marker is available.
+                for (const marker of markers) {
+                    // Break early if an error marker is present, since it represents the highest severity.
+                    if (marker.data.severity === DiagnosticSeverity.Error) {
+                        maxSeverity = marker;
+                        break;
+                    } else if (marker.data.severity === DiagnosticSeverity.Warning) {
+                        maxSeverity = marker;
+                    }
+                }
+                // Decorate the tabbar with the highest marker severity if available.
+                return maxSeverity ? [this.toDecorator(maxSeverity)] : [];
```

Comment:
now it is really optimized 👍 so before there were as many icons as we have error/warning diagnostics?

Reply:
Before, for a given `resourceUri` there would be for each marker we decorate the tabbar which would override the previous decoration (ex: error, error, error, info) and we'd end up with an info in the end since the list is unordered. Now, we only decorate once for the highest severity and only if the severity is an `error` or `warning`. I tried to optimize as much as possible :)

Reply:
I had an idea to keep here:
- a mapping between uri -> stat (count of warnings, errors, infos)
- and total stat

So when a change event happens we update a single stat by uri and then update a total stat by subtracting previous uri stat and adding new uri stat. So we can update problem stat in the status bar very fast, maybe even without any debouncing like now. It would require testing with many diagnostics though again, i.e. as in Dart extension support PR.

If we would also add an event like `onDidStatChanged`, then we could change `ProblemTabBarDecorator` based on it instead of changed to markers directly and decorate title also very fast by retrieving a uri stat and checking how many errors or warnings it has (so not traversing anymore).

Reply:
> I had an idea to keep in [ProblemManager](https://github.com/theia-ide/theia/blob/master/packages/markers/src/browser/problem/problem-manager.ts):
> 
> * a mapping between uri -> stat (count of warnings, errors, infos)
> * and total stat
> 
> So when a change event happens we update a single stat by uri and then update a total stat by subtracting previous uri stat and adding new uri stat. So we can update problem stat in the status bar very fast, maybe even without any debouncing like now. It would require testing with many diagnostics though again, i.e. as in Dart extension support PR.
> 
> If we would also add an event like `onDidStatChanged(uri)`, then we could change `ProblemTabBarDecorator` based on uri stats instead of changes to markers directly and decorate title also very fast by retrieving a uri stat and checking how many errors or warnings it has (so traversing happens only once to compute stat for a uri, not by each client on each request).
> 
> It is just thoughts, i think everything works good now already.

I think that any optimization we can do would always be worth it :)
Where I think there can be improvements is the decoration present in the explorer for example, it is probably very costly at the moment.

Reply:
@vince-fugnitto Can you file an issue for it then please?

Reply:
> @vince-fugnitto Can you file an issue for it then please?

Regarding your proposal or in general to optimize and improve decorations?

Reply:
general with a proposal in the body